packages/ground_projection/src/ground_projection_node.py

```diff
@@ -1,90 +1,287 @@
 #!/usr/bin/env python
 
-from cv_bridge import CvBridge, CvBridgeError
-from duckietown_msgs.msg import (Segment, SegmentList)
-import duckietown_utils as dtu
-from ground_projection.ground_projection_interface import GroundProjection, \
-    get_ground_projection_geometry_for_robot
-from ground_projection.srv import EstimateHomography, EstimateHomographyResponse, GetGroundCoord, GetGroundCoordResponse, GetImageCoord, GetImageCoordResponse  #@UnresolvedImport
+import cv2
+from cv_bridge import CvBridge
+import os
+import yaml
+import numpy as np
+
 import rospy
-from sensor_msgs.msg import (Image, CameraInfo)
+from duckietown.dtros import DTROS, NodeType, TopicType
+from ground_projection import Point, GroundProjectionGeometry
+from image_geometry import PinholeCameraModel
+
+from sensor_msgs.msg import CameraInfo, CompressedImage
+from geometry_msgs.msg import Point as PointMsg
+from duckietown_msgs.msg import Segment, SegmentList
+
+
+class GroundProjectionNode(DTROS):
+    """
+    This node projects the line segments detected in the image to the ground plane and in the robot's reference frame.
+    In this way it enables lane localization in the 2D ground plane. This projection is performed using the homography
+    matrix obtained from the extrinsic calibration procedure.
+
+    Args:
+        node_name (:obj:`str`): a unique, descriptive name for the node that ROS will use
 
+    Subscribers:
+        ~camera_info (:obj:`sensor_msgs.msg.CameraInfo`): Intrinsic properties of the camera. Needed for rectifying the segments.
+        ~lineseglist_in (:obj:`duckietown_msgs.msg.SegmentList`): Line segments in pixel space from unrectified images
 
-class GroundProjectionNode(object):
+    Publishers:
+        ~lineseglist_out (:obj:`duckietown_msgs.msg.SegmentList`): Line segments in the ground plane relative to the robot origin
+        ~debug/ground_projection_image/compressed (:obj:`sensor_msgs.msg.CompressedImage`): Debug image that shows the robot relative to the projected segments. Useful to check if the extrinsic calibration is accurate.
```

Comment:
I am never able to see anything on this output.

Reply:
I think it's caused by this error:
```
[ERROR] [1592339452.245835]: bad callback: <bound method DTSubscriber._monitored_callback of <duckietown.dtros.dtsubscriber.DTSubscriber object at 0x7f1344374950>>
Traceback (most recent call last):
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "/duckietown/catkin_ws/src/dt-ros-commons/packages/duckietown/include/duckietown/dtros/dtsubscriber.py", line 125, in _monitored_callback
    out = self._user_callback(*args, **kwargs)
  File "/duckietown/catkin_ws/src/dt-core/packages/ground_projection/src/ground_projection_node.py", line 140, in lineseglist_cb
    debug_image_msg = self.bridge.cv2_to_compressed_imgmsg(self.debug_image(seglist_out))
  File "/duckietown/catkin_ws/src/dt-core/packages/ground_projection/src/ground_projection_node.py", line 276, in debug_image
    pt1=(int(segment.points[0].y * -400) + 200, int(segment.points[0].x * -400) + 300),
ValueError: cannot convert float NaN to integer
```

Reply:
This was actually caused by an error in the the way the simulator was linked to ROS. It was not publishing valid camera_info's and I think that the camera_info is now being used more explicitly inside the ground_projection. img_width and height were 0 causing 'nan's. We could test for this (kind of a good idea whenever you are dividing by something)